userapi/storage/postgres/storage.go

```diff
@@ -47,14 +49,23 @@ type Database struct {
 	threepids             threepidStatements
 	openIDTokens          tokenStatements
 	keyBackupVersions     keyBackupVersionStatements
+	devices               devicesStatements
+	loginTokens           loginTokenStatements
+	loginTokenLifetime    time.Duration
 	keyBackups            keyBackupStatements
 	serverName            gomatrixserverlib.ServerName
 	bcryptCost            int
 	openIDTokenLifetimeMS int64
 }
 
+const (
+	// The length of generated device IDs
+	deviceIDByteLength   = 6
+	loginTokenByteLength = 32
```

Comment:
Is it possible to merge `storage/postgres/storage.go` and `storage/sqlite/storage.go` into a `storage/shared/storage.go` like we do elsewhere? There's a lot of logic here which is duplicated which is sadness making.

Reply:
Yep, I'll do so in a follow up PR though to keep it easier to follow.

Reply:
⭐